### PR TITLE
[libc++] Fix inconsistency between is_lock_free and is_always_lock_free

### PR DESCRIPTION
**Pull request by ldionne**
std::atomic is implemented with the following (confusing!) hierarchy of types:

     std::atomic<T> : std::__atomic_base<T> { ... };
     std::__atomic_base<T> {
          std::__cxx_atomic_impl<T> __impl;
     };
     std::__cxx_atomic_impl<T> {
          _Atomic(T) __val;
     };

Inside std::__atomic_base, we implement the is_lock_free() and is_always_lock_free() functions. However, we used to implement them inconsistently:
- is_always_lock_free() is based on whether __cxx_atomic_impl<T> is always lock free (using the builtin), which means that we include any potential padding added by _Atomic(T) into the determination.
- is_lock_free() was based on whether T is lock free (using the builtin), which meant that we did not take into account any potential padding added by _Atomic(T).

It is important to note that the padding added by _Atomic(T) can turn a type that wouldn't be lock free into a lock free type, for example by making its size become a power of two.

The inconsistency of how the two functions were implemented could lead to cases where is_always_lock_free() would return true, but is_lock_free() would then return false. This is the case for example of the following type, which is always lock free on arm64 but was incorrectly reported as !is_lock_free() before this patch:

     struct Foo { float x[3]; };

This patch switches the determination of is_lock_free() to be based on __cxx_atomic_impl<T> instead to match how we determine is_always_lock_free().

rdar://115324353

### CODE REVIEW

File: libcxx/test/std/atomics/atomics.types.operations/atomics.types.operations.req/atomic_is_lock_free.pass.cpp [hunk 1]
```diff
@@ -27,8 +27,12 @@ struct TestFn {
   void operator()() const {
     typedef std::atomic<T> A;
     T t = T();
+
```
Comment (phyBrackets) [inline libcxx/test/std/atomics/atomics.types.operations/atomics.types.operations.req/atomic_is_lock_free.pass.cpp:30]: unrelated?

**Review by phyBrackets — COMMENTED** [review]
Wow, thanks for noticing the inconsistency, fix looks good to me, as `is_lock_free` now correctly  considering padding introduced by _Atomic(T).

File: libcxx/test/std/atomics/atomics.types.operations/atomics.types.operations.req/atomic_is_lock_free.pass.cpp (hunk 1 above)
Comment (ldionne) [reply libcxx/test/std/atomics/atomics.types.operations/atomics.types.operations.req/atomic_is_lock_free.pass.cpp:30]: It made the test a bit clearer since we are testing for non-volatile and volatile symmetrically.

**Review by phyBrackets — APPROVED** [review]

**Comment by EricWF** [comment]
I'm a bit confused. Is it `_Atomic` that's adding the padding, or the struct we wrap it in?

File: libcxx/test/std/atomics/atomics.lockfree/isalwayslockfree.pass.cpp [hunk 2]
```diff
@@ -21,7 +21,6 @@
 template <typename T>
 void checkAlwaysLockFree() {
   if (std::atomic<T>::is_always_lock_free) {
-    LIBCPP_ASSERT(sizeof(std::atomic<T>) == sizeof(T)); // technically not required, but libc++ does it that way
```
Comment (EricWF) [inline libcxx/test/std/atomics/atomics.lockfree/isalwayslockfree.pass.cpp:24]: I assume the FloatArr3 case is what causes this?
Under what architecture? I haven't been able to reproduce the size difference in godbolt.

Comment (ldionne) [reply libcxx/test/std/atomics/atomics.lockfree/isalwayslockfree.pass.cpp:24]: Yes, exactly, `FloatArr3` showcases this under `arm64-apple-macos` (and probably any `arm64` target). I looked at the assembly and we do not use a lock for `FloatArr3` on arm64 so it looks like `is_lock_free` was really lying.

**Comment by ldionne** [comment]
> I'm a bit confused. Is it `_Atomic` that's adding the padding, or the struct we wrap it in?

My understanding is that it's `_Atomic(T)` itself. I think it does that to allow `Float3Arr` to be lockfree (it needs it to be a power-of-2 alignment). So `sizeof(_Atomic(Float3)) == 16` but `sizeof(Float3) == 12`.

**Review by huixie90 — APPROVED** [review]
LGTM!

**Comment by ldionne** [comment]
The CI failure is the bootstrapping build agent dying, definitely not related to this. Shipping.
